engine/src/main/java/org/terasology/rendering/animation/Animation.java

```diff
@@ -0,0 +1,142 @@
+import java.util.Collections;
+import java.util.Set;
+import java.util.List;
+import java.util.ArrayList;
+
+/*
+ */
+public class Animation implements Component, AssetData {
+    private List<AnimationListener> listeners;
+    private List<Frame> frames;
+    private int repeat;
+
+    private float elapsedTime;
+    private int currentFrame;
+    private int currentRepeatCount;
+
+    private enum AnimState {
+        PRESTART, PAUSED, RUNNING, FINISHED
+    }
+    private AnimState currentState;
+
+    public Animation() {
+        listeners = new ArrayList<AnimationListener>();
+        frames = new ArrayList<Frame>();
+        currentState = AnimState.PRESTART;
+        repeat = 0;
+        elapsedTime = 0;
+        currentFrame = 0;
+    }
+
+    public void addFrame(Frame frame) {
+        this.frames.add(frame);
+    }
+
+    public void setRepeat(int repeat) {
+        this.repeat = repeat;
+    }
+
+    public void update(float delta) {
+        switch (this.currentState) {
+        case RUNNING: {
+            frames.get(currentFrame).update(delta);
+            if (frames.get(currentFrame).isFinished()) {
+                currentFrame++;
+                if (currentFrame >= frames.size()) {
+                    currentRepeatCount++;
+                    if (currentRepeatCount >= repeat ||
+                        currentRepeatCount >= -(repeat + 1)) {
+                        end();
+                    } else { // if repeating forever or still able to repeat
+                        currentFrame = 0;
+                        elapsedTime = 0;//startDelay;
+                    }
+                }
+            }
+            break;
+        }
+        default: break;
+        }
+    }
+
+    public void start() {
+        if (this.currentState.equals(AnimState.PRESTART)) {
+            elapsedTime = 0;
+            currentRepeatCount = 0;
+            this.currentState = AnimState.RUNNING;
+            for (AnimationListener li : this.listeners) {
+                li.onStart();
+            }
+        }
+    }
+
+    public void end() {
+        if (this.currentState.equals(AnimState.RUNNING)) {
+            this.currentState = AnimState.FINISHED;
+            for (AnimationListener li : this.listeners) {
+                li.onEnd(this.currentRepeatCount);
+            }
+        }
+    }
+
+    public void pause() {
+        if (this.currentState.equals(AnimState.RUNNING)) {
+            this.currentState = AnimState.PAUSED;
+        }
+    }
+
+    public void resume() {
+        if (this.currentState.equals(AnimState.PAUSED)) {
+            this.currentState = AnimState.RUNNING;
+        }
+    }
+
+    public void addListener(AnimationListener li) {
+        this.listeners.add(li);
+    }
+
+    public void removeListener(AnimationListener li) {
+        this.listeners.remove(li);
+    }
+
+    public boolean isFinished() {
+        return currentState.equals(AnimState.FINISHED);
+    }
+
+    public interface AnimationListener {
```

Comment:
I haven't done anything with default methods in interfaces, I'll have to do some research.